netket/optimizer/qgt/qgt_jacobian_pytree.py

```diff
@@ -111,27 +183,13 @@ def to_dense(self) -> jnp.ndarray:
 
         Returns:
             A dense matrix representation of this S matrix.
+            In R→R and R→C modes, real and imaginary parts of parameters get own rows/columns
         """
-        if scale is None:
-            O = self.O
-            diag = jnp.eye(self.O.shape[1])
+        if self.mode == "holomorphic":
+            pars = self.params
         else:
-            O = self.O * self.scale[jnp.newaxis, :]
-            diag = jnp.diag(self.scale ** 2)
+            pars = nkjax.tree_to_real(self.params)
 
-        return sum_inplace(O.T.conj() @ O) + self.diag_shift * diag
-
-
-def _grad_vmap_minus_mean(
-    fun: Callable, params: jnp.ndarray, samples: jnp.ndarray, holomorphic: bool
-):
-    """Calculates the gradient of a neural network for a number of samples
-    efficiently using vmap(grad),
-    and subtracts their mean for each parameter, i.e., each column
-    """
-    grads = jax.vmap(
-        jax.grad(fun, holomorphic=holomorphic), in_axes=(None, 0), out_axes=0
-    )(params, samples)
-    return grads - sum_inplace(grads.sum(axis=0, keepdims=True)) / (
-        grads.shape[0] * n_nodes
-    )
+        Npars = nkjax.tree_size(pars)
+        I = jax.numpy.eye(Npars)
+        return jax.vmap(self._split_matmul, in_axes=0)(I)
```

Comment:
@attila-i-szabo I think this contraption is meant, which is borrowed from the onthefly code. Here we should just do dense matmul

Reply:
I see. I wasn't thinking carefully about the matmul in your logic code, does that support dense products?

Reply:
although the biggest difficulty would be explaining the `_split_matmul` routine how to turn an arbitrary wide matrix into the right pytree. or rather, I could do that with `vmap`, but then what do we gain?